content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,97 @@
+. *HttpRequest:* Many times this would be used to grab data from an external source and store it into a variable. This is not a good idea to use because not only is that request coming directly from the master(which could give incorrect results for things like HTTPS requests if the master does not have certificates loaded) but also the return is again stored twice.
+.. Solution: Again using the shell step but this time with a `curl` command is the best direction to go.Then have the request come from master and also the data size can be shrunk with filtering.
+
+=== Reduce Repetition of similar Pipeline steps: 
+
+Combine Pipeline steps into single steps as often as possible to reduce the size of the flow-node graph(Pipeline’s way of tracking what step you are currently on). 
+This flow node graph has to be loaded onto the stack and if it is very large it can end up causing memory issues or even a StackOverflow.
+
+Example:
+Whenever doing a series of `echo` or `sh` steps make sure to combine them into a single step or script rather than individual steps
+
+=== Avoid Calling `Jenkins.getInstance`
+
+Using Jenkins.instance or its accessor methods in a Pipeline or shared library indicates a code misuse within that Pipeline/shared library. Using Jenkins APIs from an unsandboxed shared library means that the shared library is both a shared library and a kind of Jenkins plugin. You need to be very careful when interacting with Jenkins APIs from a Pipeline to avoid severe security and performance issues. If you must use Jenkins APIs in your build, the recommend approach is to create a minimal plugin in Java that implements a safe wrapper around the Jenkins API you want to access using Pipeline's Step API. Using Jenkins APIs from a sandboxed Jenkinsfile directly means that you have probably had to whitelist methods that allow sandbox protections to be bypassed by anyone who can modify a Pipeline, which is a significant security risk. The whitelisted method is run as the System user meaning it has overall admin permissions and can lead to developers running with higher permissions than intended(causing the security risk).
+
+Solution: The best solution would be to work around the calls being made but if they must be done then it would be better to implement a Jenkins plugin which is able to gather the data needed.
+
+== Shared Libraries
+
+=== Just say no to overriding built-in Pipeline Steps
+
+Wherever possible stay away from customized/overwritten Pipeline steps. 
+This is the process where the standard Pipeline APIs like `sh` or `timeout` are overwritten using shared libraries. 
+It is dangerous to do this because the Pipeline APIs can change at any time causing custom code to break or give different results than expected. 
+These changes can also become difficult to troubleshoot when something is breaking or causing issues. 
```

Comment:
When custom code breaks because of Pipeline API changes, troubleshooting is difficult because even if the custom code has not changed, it may not work the same after an API update.